app/main/browser.js

```diff
@@ -48,24 +47,41 @@ const globalShortcut = electron.globalShortcut;
 
 const uiDir = path.join(__dirname, '..', 'ui');
 
-// Keep a global references of the window objects, if you don't, the windows will
-// be closed automatically when the JavaScript object is garbage collected.
-const mainWindows = [];
+const store = configureStore();
+let currentState;
 
-function sendToAllWindows(event, args) {
+function sendToAllWindows(event: string, args: Object): void {
   console.log(`diff ${JSON.stringify(args)}`);
-  for (const mainWindow of mainWindows) {
-    mainWindow.webContents.send(event, args);
-  }
+  store.getState().browserWindows.valueSeq().forEach((bw) => {
+    console.log(`send to ${bw.id}`);
+    bw.webContents.send(event, args);
+  });
 }
 
-const store = configureStore();
-let currentState;
-
-function sendDiffsToWindows() {
+function sendDiffsToWindows(): void {
   const previousState = currentState;
   currentState = store.getState();
 
+  // TODO: handle empty state.
+  if (previousState && !Immutable.is(currentState.browserWindows.keySeq(),
+                                     previousState.browserWindows.keySeq())) {
+    // Show new windows, taking care to use key IDs rather than possibly deleted ID members.
+    currentState.browserWindows.forEach((bw, id) => {
```

Comment:
We should probably yield on the showing and closing of windows for the entire function, but still run in parallel.

```js
let opening = browserWindows.map((bw,id) => currentState.has(id) ? null : bw.p.then(()=>bw.show()));
let closing = browserWindows.map((bw,id) => prevState.has(id) ? null : bw.p.then(()=>bw.close()));

yield Promise.all([opening, closing])
```